### PR TITLE
#10080 Convert CuDNN weights in nested Model.

### PR DESCRIPTION
**Pull request by bzamecnik**
- similar problem to nesting in Bidirectional (#8860)
- ~quick fix: just copy/paste. I'll refactor it later~
- also convert biases from H5 Dataset to np.array for reshape() (#9662)
- [x] refactor duplicate code
- [x] add tests
  - nested models
  - various model types: Model, Sequential (not only Sequential)
    - [x] Model -> CuDNNGRU
    - [x] Sequential -> CuDNNGRU
    - [x] Model -> Model -> CuDNNGRU
    - ~Model -> Sequential -> CuDNNGRU~
    - ~Sequential -> Model -> CuDNNGRU~
    - [x] Sequential -> Sequential -> CuDNNGRU
  - [x] convert weights on model, not layer
  - [x] better just save model weights and load them back instead of calling `preprocess_weights_for_loading()` directly
- fixed `pytest.skipif` markers for requiring TensorFlow and GPU
  - It seems that pytest doesn't support short-circuit evaluation of subsequent markers. I can't understand how the tests passed until now.
- refactored duplicate `pytest.skipif` markers

### CODE REVIEW

File: keras/engine/saving.py [hunk 1]
```diff
@@ -491,7 +492,12 @@ def preprocess_weights_for_loading(layer, weights,
     # Returns
         A list of weights values (Numpy arrays).
     """
-    if layer.__class__.__name__ == 'Bidirectional':
+    def convert_nested_bidirectional(weights):
+        """
+        Converts layers nested in `Bidirectional` wrapper by `preprocess_weights_for_loading()`.
+        :param weights: list of of numpy arrays
+        :return: converted list of of numpy arrays
+        """
```
Comment (fchollet) [inline keras/engine/saving.py]: Please format the docstring like other docstrings in the codebase.

File: keras/engine/saving.py [hunk 2]
```diff
@@ -477,7 +477,8 @@ def preprocess_weights_for_loading(layer, weights,
                                    original_keras_version=None,
                                    original_backend=None,
                                    reshape=False):
-    """Converts layers weights from Keras 1 format to Keras 2.
+    """Converts layers weights from Keras 1 format to Keras 2
+    and also weights of CuDNN layers in Keras 2.
```
Comment (fchollet) [inline keras/engine/saving.py]: Docstring summary should fit on one line

File: keras/engine/saving.py [hunk 3]
```diff
@@ -501,7 +507,46 @@ def preprocess_weights_for_loading(layer, weights,
...
                                                           original_keras_version,
                                                           original_backend)
-        weights = forward_weights + backward_weights
+        return forward_weights + backward_weights
+
+    def convert_nested_model(weights):
+        """
+        Converts layers nested in `Model` or `Sequential` by `preprocess_weights_for_loading()`.
+        :param weights: list of of numpy arrays
+        :return: converted list of of numpy arrays
```
Comment (fchollet) [inline keras/engine/saving.py]: Same comment about docstring format

File: keras/engine/saving.py [hunk 4]
```diff
@@ -501,7 +507,46 @@ def preprocess_weights_for_loading(layer, weights,
...
+        """
+        new_weights = []
+        # trainable weights
+        for sublayer in layer.layers:
+            num_weights = len(sublayer.trainable_weights)
+            if num_weights > 0:
```
Comment (fchollet) [inline keras/engine/saving.py]: You can use `if sublayer.trainable_weights`

File: keras/engine/saving.py [hunk 5]
```diff
@@ -501,7 +507,46 @@ def preprocess_weights_for_loading(layer, weights,
...
+                new_weights.extend(preprocess_weights_for_loading(
+                    layer=sublayer,
+                    weights=weights[:num_weights],
+                    original_keras_version=original_keras_version,
+                    original_backend=original_backend))
+                weights = weights[num_weights:]
+
+        # non-trainable weights
+        for sublayer in layer.layers:
+            num_weights = len([l for l in sublayer.weights
```
Comment (fchollet) [inline keras/engine/saving.py]: Weights are either trainable or not. Why is this necessary?

File: keras/engine/saving.py [hunk 6]
```diff
@@ -501,7 +507,46 @@ def preprocess_weights_for_loading(layer, weights,
...
+        # trainable weights
+        for sublayer in layer.layers:
+            num_weights = len(sublayer.trainable_weights)
+            if num_weights > 0:
+                new_weights.extend(preprocess_weights_for_loading(
+                    layer=sublayer,
+                    weights=weights[:num_weights],
+                    original_keras_version=original_keras_version,
+                    original_backend=original_backend))
+                weights = weights[num_weights:]
```
Comment (fchollet) [inline keras/engine/saving.py]: Somewhat confusing, better to select the correct slice in the paragraph below

File: keras/engine/saving.py (hunk 1 above)
Comment (bzamecnik) [reply keras/engine/saving.py]: Thanks.

File: keras/engine/saving.py (hunk 5 above)
Comment (bzamecnik) [reply keras/engine/saving.py]: Well, I just moved code from one place to another. I'm not sure why somebody wrote it like this. IMHO refactoring that code is for another PR.

Comment (fchollet) [reply keras/engine/saving.py]: Ok, SG

**Comment by fchollet** [comment]
Please fix PEP8 issue: https://travis-ci.org/keras-team/keras/builds/374509126?utm_source=github_status&utm_medium=notification

**Comment by bzamecnik** [comment]
Thanks! 🎉
